File: frontend/tests/ui/test-login.js
// UI 自动化测试 - 登录功能（增强版）
const { chromium } = require('playwright');

async function testLogin() {
    console.log('1. 启动浏览器...');
    const browser = await chromium.launch({
        headless: true,
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage', '--disable-gpu']
    });
    const context = await browser.newContext();
    const page = await context.newPage();

    // 监听控制台消息
    page.on('console', msg => console.log('   [浏览器]', msg.text()));

    // 监听页面错误
    page.on('pageerror', error => console.error('   [页面错误]', error.message));

    // 监听请求失败
    page.on('requestfailed', request => {
        console.error('   [请求失败]', request.url(), request.failure().errorText);
    });

    try {
        console.log('2. 访问登录页面...');
        await page.goto('http://localhost:5174', { timeout: 30000 });
        await page.waitForLoadState('networkidle', { timeout: 30000 });
        console.log('   ✓ 页面加载完成');

        console.log('3. 检查登录表单...');
        const usernameInput = page.locator('input[placeholder*="username" i]');
        const passwordInput = page.locator('input[type="password"]');
        const loginButton = page.locator('button[type="submit"]');

        await usernameInput.waitFor({ state: 'visible', timeout: 10000 });
        await passwordInput.waitFor({ state: 'visible', timeout: 10000 });
        await loginButton.waitFor({ state: 'visible', timeout: 10000 });
        console.log('   ✓ 登录表单元素已找到');

        console.log('4. 验证自动填充...');
        const usernameValue = await usernameInput.inputValue();
        const passwordValue = await passwordInput.inputValue();

        console.log('   用户名:', usernameValue);
        console.log('   密码:', '*'.repeat(passwordValue.length));

        if (usernameValue !== 'admin') {
            throw new Error('用户名应为 admin，实际为 ' + usernameValue);
        }
        if (passwordValue !== 'admin666') {
            throw new Error('密码应为 admin666');
        }
        console.log('   ✓ 自动填充验证通过');

        console.log('5. 点击登录按钮...');

        // 截图：登录前
        await page.screenshot({ path: '/tmp/before_login.png' });
        console.log('   ✓ 登录前截图: /tmp/before_login.png');

        // 等待导航
        const navigationPromise = page.waitForURL(/localhost:5174/, { timeout: 30000 });
        await loginButton.click();

        console.log('6. 等待登录完成...');

        try {
            await navigationPromise;
            console.log('   ✓ 页面已跳转');
        } catch (e) {
            console.log('   ! 等待跳转超时，检查当前状态...');
        }

        // 等待一下让页面稳定
        await page.waitForTimeout(3000);

        // 截图：登录后
        await page.screenshot({ path: '/tmp/after_login.png' });
        console.log('   ✓ 登录后截图: /tmp/after_login.png');

        console.log('7. 验证登录状态...');
        const currentUrl = page.url();
        console.log('   当前 URL:', currentUrl);

        // 检查是否有错误提示
        const errorMessage = await page.locator('.el-message--error').count();
        if (errorMessage > 0) {
            const errorText = await page.locator('.el-message--error').first().textContent();
            throw new Error('登录失败: ' + errorText);
        }

        // 检查是否还在登录页面
        if (currentUrl.includes('login')) {
            throw new Error('登录失败，仍在登录页面');
        }

        // 检查是否有导航菜单（说明已登录）
        const hasNav = await page.locator('.el-menu, nav, .sidebar').count() > 0;
        if (hasNav) {
            console.log('   ✓ 检测到导航菜单，登录成功');
        } else {
            console.log('   ! 未检测到导航菜单，但已离开登录页');
        }

        console.log('\n✅ 登录测试通过！');
        console.log('   最终 URL:', currentUrl);

    } catch (error) {
        console.error('\n❌ 测试失败:', error.message);

        // 失败时的详细信息
        try {
            const currentUrl = page.url();
            console.error('   当前 URL:', currentUrl);

            // 截图
            await page.screenshot({ path: '/tmp/test_failed.png', fullPage: true });
            console.error('   失败截图: /tmp/test_failed.png');

            // 获取页面内容
            const bodyText = await page.locator('body').textContent();
            console.error('   页面内容（前200字符）:', bodyText.substring(0, 200));
        } catch (e) {
            console.error('   无法获取失败详情:', e.message);
        }

        throw error;
    } finally {
        await browser.close();
    }
}

async function main() {
    console.log('='.repeat(60));
    console.log('UI 自动化测试 - 登录功能（增强版）');
    console.log('='.repeat(60));

    try {
        await testLogin();
        console.log('\n' + '='.repeat(60));
        console.log('所有测试通过！✅');
        console.log('='.repeat(60));
    } catch (error) {
        console.error('\n' + '='.repeat(60));
        console.error('测试失败 ❌');
        console.error('='.repeat(60));
        process.exit(1);
    }
}

main();

File: frontend/tests/ui/test_login.py
"""
UI 自动化测试 - 登录功能

使用 Playwright 测试登录流程
"""
import asyncio
from playwright.async_api import async_playwright, expect


async def test_login():
    """测试登录功能"""
    async with async_playwright() as p:
        # 启动浏览器
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context()
        page = await context.new_page()
        
        print("1. 访问登录页面...")
        await page.goto("http://localhost:5174")
        
        # 等待页面加载
        await page.wait_for_load_state("networkidle")
        
        print("2. 检查登录表单...")
        # 检查用户名输入框
        username_input = page.locator('input[placeholder*="username"]')
        await expect(username_input).to_be_visible()
        
        # 检查密码输入框
        password_input = page.locator('input[type="password"]')
        await expect(password_input).to_be_visible()
        
        print("3. 验证自动填充...")
        # 验证用户名已自动填充
        username_value = await username_input.input_value()
        print(f"   用户名: {username_value}")
        assert username_value == "admin", f"用户名应为 'admin'，实际为 '{username_value}'"
        
        # 验证密码已自动填充
        password_value = await password_input.input_value()
        print(f"   密码: {'*' * len(password_value)}")
        assert password_value == "admin666", "密码应为 'admin666'"
        
        print("4. 点击登录按钮...")
        # 查找并点击登录按钮
        login_button = page.locator('button[type="submit"]')
        await expect(login_button).to_be_visible()
        await login_button.click()
        
        print("5. 等待登录完成...")
        # 等待页面跳转
        await page.wait_for_url("http://localhost:5174/", timeout=10000)
        
        print("6. 验证登录成功...")
        # 检查是否跳转到主页
        current_url = page.url
        print(f"   当前 URL: {current_url}")
        assert current_url == "http://localhost:5174/" or current_url.startswith("http://localhost:5174/dashboard"), \
            f"登录后应跳转到主页，实际 URL: {current_url}"
        
        # 检查是否有导航菜单或用户信息
        # 等待主页面元素加载
        await page.wait_for_timeout(2000)
        
        # 截图保存
        await page.screenshot(path="/tmp/login_success.png")
        print("   ✓ 登录成功截图已保存: /tmp/login_success.png")
        
        print("\n✅ 登录测试通过！")
        
        await browser.close()


async def test_login_with_wrong_password():
    """测试错误密码登录"""
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context()
        page = await context.new_page()
        
        print("\n测试错误密码...")
        await page.goto("http://localhost:5174")
        await page.wait_for_load_state("networkidle")
        
        # 清空并输入错误密码
        password_input = page.locator('input[type="password"]')
        await password_input.fill("wrong_password")
        
        # 点击登录
        login_button = page.locator('button[type="submit"]')
        await login_button.click()
        
        # 等待错误提示
        await page.wait_for_timeout(2000)
        
        # 检查是否显示错误信息
        error_message = page.locator('.el-message--error')
        if await error_message.count() > 0:
            print("   ✓ 显示了错误提示")
        
        # 验证仍在登录页面
        current_url = page.url
        assert "login" in current_url or current_url == "http://localhost:5174/", \
            "错误密码应该停留在登录页面"
        
        print("   ✓ 错误密码测试通过")
        
        await browser.close()


async def main():
    """运行所有测试"""
    print("=" * 60)
    print("UI 自动化测试 - 登录功能")
    print("=" * 60)
    
    try:
        await test_login()
        await test_login_with_wrong_password()
        
        print("\n" + "=" * 60)
        print("所有测试通过！✅")
        print("=" * 60)
    except Exception as e:
        print(f"\n❌ 测试失败: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())
